Fix/clientside sagas

zentuit opened the pull request:
Update `withReduxSaga` to work on client side.

Change `withReduxSaga` API to take either an object with `server` and `client` fields identifying respective action types.  For example, passing
```
export default withReduxSaga(Home, {
  server: { type: 'GET_REPO_SERVER' },
  client: { type: 'GET_REPO' },
});
```
will have the action type 'GET_REPO_SERVER' dispatched on the server while 'GET_REPO' is dispatched on the client side by `getInitialProps`.

I also dispatch the entire context received by `getInitialProps` to allow any watching sagas full access.

I also added hot module reloading for sagas. Now editing a saga gives the HMR a chance to reload sagas.

Commit:
fix client side sagas

The client side actions were never fired so update the withReduxSaga api
to accept an object that has server/client keys with respective action
types so the correct action is called in getInitialProps

Commit:
update movieDetail component to display more info

Commit:
remove some debug logging

Commit:
remove rootSagas import

runSagas can figure out which sagas to run by itself so we can
encapsulate that a bit

batusai513 commented:
@zentuit Thank you for this PR, I'll check it tonight 👍

zentuit commented:
@batusai513 : you may want to hold off on checking it as I have another update in the works.

ReLrO commented:
Hi @zentuit Did you ever finish working on that update?

Commit:
update to new `withReduxSaga`

I've refactored and reworked `withReduxSaga`, `store/index.js` and the
various saga files to come up with a way to do `withReduxSaga` as a
possible npm package.

Everything needed for `withReduxSaga` to work is confined to the
`withReduxSaga` directory.

The `store/index.js` needed to change the export from a function to an
object that holds the `configure` function and the array of `rootSagas`.
The `rootSagas` are used internally by `withReduxSaga`.

zentuit commented:
@ReLrO : now that's timing!  I just finished cleaning things up to push in this branch.

@batusai513 : here's the update...

Its a big change.  My goal was to make it so that `withReduxSaga` could be published as an npm package. I think I accomplished that goal.

To reach that goal, a number of things changed including the API for `withReduxSaga`.  Before the `page` component was written like so:
```
import React from 'react';
import withReduxSaga from '../withReduxSaga';
import MainLayout from '../components/mainLayout';
import Repo from '../components/repo';

function Home() {
  return (
    <MainLayout>
      <Repo />
    </MainLayout>
  );
}

export default withReduxSaga(Home, { type: 'GET_REPO' });
```
now its written as so:
```
import React from 'react';
import withReduxSaga from '../withReduxSaga';
import MainLayout from '../components/mainLayout';
import Repo from '../components/repo';
import configureStore from '../store';

function Home() {
  return (
    <MainLayout>
      <Repo />
    </MainLayout>
  );
}

export default withReduxSaga(configureStore, Home, { type: 'GET_REPO' });
```
(note that the action parameter can still identify separate actions for client and server (see `/pages/movie.js`)

The next thing that changed is how `configureStore` is exported. Now its an object :
```
export default {
  configureStore,
  rootSaga,
  sagaMonitor,
};
```
The `configureStore` is a function, the `rootSaga` is an array of sagas, and the `sagaMonitor` is an optional entry which will let you hook in the (`sagaMonitor`)[https://github.com/clarketm/saga-monitor].

I have to run to a meeting right now. If you all would like to pull it and play around with it, I'd like your feedback.

batusai513 commented on withReduxSaga/withReduxSaga.js:
```diff
@@ -0,0 +1,31 @@
+    }
+  }
+
+  return withRedux(params.makeStore)(ReduxContainer);
```
the first function param should be `params.makeStore.createStore`

zentuit replied:
@batusai513 : the params.makeStore is correct. `withRedux` takes either a function or an configuration object. If its an object, then there has to be a field `createStore` and there are a few other options - the one I need is `storeKey`. This sets the redux store key to something that I can control and so can lookup in `window` object to see if the store exists already.

batusai513 replied:
when i was testing it was trowing an error because of that :/, probably a bad pull on my side?

zentuit replied:
@batusai513 : was this project using yarn or npm?  My team member was getting an error when he tested it because he was using yarn and I (maybe mistakenly) was using npm5.

We usually use yarn but I though that his project was using npm.

zentuit replied:
@batusai513 : I just looked at master and yes that's most likely the issue.  Sorry about that, most of our projects are using yarn but a new project just started using npm5 and I got confused thinking this project was using npm5.

I can see about pushing up an update that doesn't use npm5 and instead uses yarn.

zentuit commented:
@batusai513 / @ReLrO : anyone get a chance to play around with this? Everything is working smoothly here, but I'd like to get feedback on different users/setups.

batusai513 reviewed (commented):
Oops, i forgot to submit the review lol

batusai513 commented on pages/index.js:
```diff
@@ -11,4 +12,4 @@ function Home() {
 }
 
-export default withReduxSaga(Home, { type: 'GET_REPO_SERVER' });
+export default withReduxSaga(configureStore, Home, { type: 'GET_REPO' });
```
It would be awesome to make this closer to redux api, return a function that accepts the component, a lot of people is used to do that.

zentuit replied:
So something like:
`withReduxSaga(configureStore)(Home, { type: 'GET_REPO' });` ?

That is absolutely doable!

batusai513 replied:
something like `withReduxSaga(configureStore, actions)(Component)`, sorry for not adding an example before

zentuit replied:
I think I should expand this API to mimic `withRedux`.  that means

`withReduxSaga(configObject, mapStateToProps, mapDispatchToProps, mergeProps, connectOptions)(Component, actionObject)`. 

`mapStateToProps`, `mapDispatchToProps`, and `mergeProps` are optional just like using `withRedux` (or `connect`).

I need to think about this.

batusai513 replied:
Let me know if you need some help with this, I'll gladly help you.

batusai513 commented on withReduxSaga/index.js:
```diff
@@ -0,0 +1,45 @@
+  };
+  params.makeStore = {
+    createStore: (initialState, options) => {
+      params.store = configureStore(initialState, options);
```
i've been trying to thing about how do we add the initialState easily?

zentuit replied:
`params.makeStore.createStore` is called by `withRedux` with the `initialState` it has built up. I intercept that store building so I can get the reference to the store. This does give us a side effect of having an opportunity to add additional state to the `configureStore` function. I'm just not sure what we would add.

I guess I could add an optional parameter that identifies some initial state. My gut feeling however is to let the saga handle setting default state.

batusai513 replied:
I made some emphasis on that because i tend to use a file where i build all my initial state.
```js
export default {
  entities: {
    exercises: {},
    frequencies: {},
    goals: {},
    providers: {},
    recipients: {},
    messages: {},
    conversations: {},
  },
  exercises: {
    todaysExercises: {
      isFetching: false,
      list: [],
      error: '',
    },
    library: {
      isFetching: false,
      list: [],
      error: '',
    },
  },
  goals: {
    isFetching: false,
    list: [],
    error: '',
  },
  conversations: {
    isFetching: false,
    isCreating: false,
    lastFetched: '',
    list: [],
    error: '',
    conversation: {
      id: '',
      isFetching: false,
    },
  },
  providers: {
    isFetching: false,
    lastFetched: '',
    list: [],
    error: '',
  },
};
```

zentuit replied:
I missed this response, sorry.

I'll think about how to do this cleanly.

I'm also wondering if the withReduxSaga should call the wrapped component's `getInitialState` and return it.  Which brings up the next question, does it call it before or after the store dispatch?  Things to think about.

Commit:
modify API to make it more like `withRedux`

Commit:
move mapStateToProps from component to page

Since `withReduxSaga` not accepts mapStateToProps and the other optional
parameters, we can move them up to the page level

zentuit commented:
@batusai513 : I updated the API based on your comments.  Then it let me move the mapStateToProps stuff up to the page level. Very cool, very clean.

Commit:
upgrade next-redux-wrapper and next packages

zentuit commented:
@batusai513 : `next-redux-wrapper` needed to be upgraded to 1.3 to pick up the configuration object api option.  I also upgraded `next` to the latest 3.x version since its been released.

batusai513 commented:
Thanks for your effort 👍  @zentuit

zentuit commented:
@batusai513 : are you good with making this a standalone npm package?  I don't mind doing it but its your project so your decision 😄
